odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+
+        If ``other`` is an operator, since a functional is also an operator
+        this corresponds to operator composition:
+
+            ``op * func <==> (x --> op(func(x))``
+
+        If ``other`` is a scalar, this corresponds to left multiplication of
+        scalars with functionals:
+
+            ``scalar * func <==> (x --> scalar * func(x))``
+
+        If ``other`` is a vector,  since a functional is also an operator this
+        corresponds to left multiplication of vectors with operators:
+
+            ``vector * func <==> (x --> vector * func(x))``
+
+        Note that left and right multiplications are generally
+        different.
+
+        Parameters
+        ----------
+        other : `Operator` or `LinearSpaceVector` or scalar
+            `Operator`:
+            The `Operator.domain` of ``other`` must match this functional's
+            `Functional.range`.
+
+            `LinearSpaceVector`:
+            ``other`` must be an element of this functionals's
+            `Functional.range`.
+
+            scalar:
+            The `Operator.domain` of this operator must be a
+            `LinearSpace` and ``other`` must be an
+            element of the ``field`` of this operator's
+            `Operator.domain`.
+
+        Returns
+        -------
+        rmul : `Functional` or `Operator`
+            Multiplication result
+
+            If ``other`` is an `Operator`, ``rmul`` is a `OperatorComp`.
+
+            If ``other`` is a scalar, ``rmul`` is a
+            `FunctionalLeftScalarMult`.
+
+            If ``other`` is a vector, ``rmul`` is a
+            `OperatorLeftVectorMult`.
+        """
+        if other in self.domain.field:
+            if other == 0:
+                from odl.solvers.functional.default_functionals import (
+                    ZeroFunctional)
+                return ZeroFunctional(self.domain)
+            else:
+                return FunctionalLeftScalarMult(self, other)
+        else:
+            return super().__rmul__(other)
+
+    def __add__(self, other):
+        """Return ``self + other``.
+
+        Parameters
+        ----------
+        other : `Functional` or scalar
+            `Functional`:
+            The `Functional.domain` and `Functional.range` of ``other``
+            must match this functional's  `Functional.domain` and
+            `Functional.range`.
+
+            scalar:
+            The scalar needs to be in this functional's `Functional.range`.
+
+        Returns
+        -------
+        add : `Functional`
+            Addition result
+
+            If ``other`` is in ``Functional.range``, ``add`` is a
+            `FunctionalScalarSum`.
+
+            If ``other`` is a `Functional`, ``add`` is a `FunctionalSum`.
+        """
+        if other in self.domain.field:
+            return FunctionalScalarSum(self, other)
+        elif isinstance(other, Functional):
+            return FunctionalSum(self, other)
+        else:
+            return NotImplemented
```

Comment:
The only thing that this would let through that do not pass already is when `Operator` is in fact a functional. What is potentially lost in this case is to for a gradient, but maybe we can live with that :)